### PR TITLE
Add support for bzlmod

### PR DESCRIPTION
**Pull request by luispadron**
Adds initial support for bzlmod to rules_ios.

Major changes:

- Update versions in `repositories.bzl` to use minimum supported bzlmod version
- Add `MODULE.bazel` and required files to enable bzlmod by default in Bazel 6+
- Enable bzlmod by default in the repository

Required workarounds:

- Disable bzlmod when running legacy .xcodeproj tests. This rule collects workspace name in an aspect which now resolves to simply `_main` in bzlmod. This doesnt seem to cause any actual issues with the generated project besides the repository being named `_main`. Proper fix is to have this use runfiles but I do not have enough context to make that change. 

Things to follow-up on:

- Add bzlmod support for stardoc
  - This requires some more changes to bzl_library usage and update to Stardoc version, will do this in a separate PR before publishing.
- Publish to Bazel Central Registry (and introduce files to required to do so)
- Finalize documentation and update README (after publication)
- Consider better support for legacy xcode project generator, the tests will continue to run without bzlmod as the aspect requires workspace name collection udpates.

Depends on: #716, #719

### CODE REVIEW

**Comment by luispadron** [comment]
@thiagohmcruz mostly done with this PR, however, Xcode tests are failing probably because of how repos are named under bzlmod (they don't have a standard name). 

Do you know where the logic is to determine the workspace name for the Xcode generator?

**Comment by thiagohmcruz** [comment]
IIUC your question, this:

https://github.com/bazel-ios/rules_ios/blob/master/rules/legacy_xcodeproj.bzl#L829

gets its value from the aspect collecting info here:

https://github.com/bazel-ios/rules_ios/blob/master/rules/legacy_xcodeproj.bzl#L265

**Comment by thiagohmcruz** [comment]
Should this add a `Bzlmod Snippet` to releases so folks know how to consume it using Bzlmod? Also we've introduced releases recently but I assume folks might continue to consume `HEAD` for some time, should we have instructions in the root `README` on how to load it if you're using Bzlmod?

**Comment by luispadron** [comment]
> IIUC your question, this:
> 
> https://github.com/bazel-ios/rules_ios/blob/master/rules/legacy_xcodeproj.bzl#L829
> 
> gets its value from the aspect collecting info here:
> 
> https://github.com/bazel-ios/rules_ios/blob/master/rules/legacy_xcodeproj.bzl#L265

Yep thats the attr we need to fixup, it's always `_main` when using `bzlmod`, this is because with bzlmod enabled `workspace_name()` and equivalent native functions in rules/aspects always return `_main`. 

From some digging the proper way to do this in bzlmod is to collect the repository name from the runfiles somehow, either via their path or some other mechanism. I don't really understand the generator enough to do that update. This continues to work in both modes, but with the bzlmod quirk of having `_main` as the Xcode project name.

I'm thinking of adding the legacy xcode tests to just the LTS CI jobs which don't use bzlmod.


> Should this add a `Bzlmod Snippet` to releases so folks know how to consume it using Bzlmod? Also we've introduced releases recently but I assume folks might continue to consume `HEAD` for some time, should we have instructions in the root `README` on how to load it if you're using Bzlmod?

Totally, I'm thinking of merging initial support which is barebones and following up with the required tools/docs to publish this into the BCR

File: .github/workflows/tests.yml
Comment (luispadron) [inline .github/workflows/tests.yml:159]: I'll create a separate PR for this it will update the version to the required stardoc (requires some changes to our use of `bzl_library`)

File: BUILD.bazel
Comment (luispadron) [inline BUILD.bazel:8]: This replaces the custom `sh_binary` wrapped implementation, additionally this supports bzlmod.

File: rules/module_extensions.bzl
Comment (luispadron) [inline rules/module_extensions.bzl:19]: This loads things not supported in the workspace (i.e. the repository rules) we had in WORKSPACE previously

File: rules/module_extensions.bzl
Comment (luispadron) [inline rules/module_extensions.bzl:24]: This loads and runs the xcode configuration that was previously in the WORKSPACE

File: rules/repositories.bzl
Comment (luispadron) [inline rules/repositories.bzl:60]: Changes here:

- Update versions to minimum required for bzlmod
- Move into one additional branch for bzlmod dependencies (i.e. things that support bzlmod)
  - With bzlmod these are loaded in `MODULE.bazel` so they need to be filtered here but with older WORKSPACE we default to include these

File: tests/xcodeproj-tests.sh
Comment (luispadron) [inline tests/xcodeproj-tests.sh:1]: Changes here are to disable bzlmod support for test purposes, this still works from what I can tell in bzlmod but has a `_main` for the workspace name (i.e. fails CI tests).  

I looked at updating and was blocked for a while, the change required seems to be to update the aspect that collects this the workspace name to not do that during aspect time and instead use the runfiles API to access the repository mappings file.

File: BUILD.bazel
Comment (jerrymarino) [inline BUILD.bazel:12]: Does this build it from source though?

File: rules/repositories.bzl
Comment (jerrymarino) [inline rules/repositories.bzl:172]: In the lines you've removed we already pulled the offical sources:
https://github.com/bazel-ios/rules_ios/pull/703/files#diff-5493ff8e9397811510e780de47c57abb70137f1afe85d1519130dc3679d60ce5L66

Unless there is blocking reason to change this, I'd prefer to keep that. Is there a reason you just update the URL to use a newer version of the official cut?

File: rules/repositories.bzl
Comment (jerrymarino) [inline rules/repositories.bzl:80]: Why do we need this `rules_ios_dev_dependencies` function - this just adds so much complexity to the setup and repo. Bazel should only download what it is using unless you start running wildcard queries which is a performance problem that this looks like you are attempting to fix

File: BUILD.bazel
Comment (luispadron) [reply BUILD.bazel:12]: `buildifier-prebuilt` just handles downloading the right `buildifier` for your architecture. It's already built

File: rules/repositories.bzl
Comment (luispadron) [reply rules/repositories.bzl:172]: The previous WORKSPACE used an `http_file` download, I thought it made sense to use the `buildifier-prebuilt` which supports switching based on platform and supports bzlmod

File: rules/repositories.bzl
Comment (luispadron) [reply rules/repositories.bzl:80]: Because in bzlmod you mark things as being development only dependencies when they aren't required by consumers of the rules. No one consuming rules_ios needs to download buildifier, which bzlmod solves. 

This specific change lets us reuse `repositories.bzl` while we transition to bzlmod completely (Bazel 9 I believe).

Comment (luispadron) [reply rules/repositories.bzl:80]: Some examples: https://github.com/MobileNativeFoundation/rules_xcodeproj/blob/main/MODULE.bazel#L30

Comment (luispadron) [reply rules/repositories.bzl:80]: > this just adds so much complexity to the setup

fwiw this shouldn't need to be run by end users, it's just a way for us internally to fetch development repos in both bzlmod / non-bzlmod workflows.

**Comment by luispadron** [comment]
Will be blocked by #773 if that merges.

File: rules/module_extensions.bzl
Comment (jerrymarino) [reply rules/module_extensions.bzl:24]: Isn't it fair to say that bzlmod was updated to work with existing repository rules? e.g. the [`use_repo`](https://github.com/bazel-ios/rules_ios/pull/703/files#diff-6136fc12446089c3db7360e923203dd114b6a1466252e71667c6791c20fe6bdcR58) wrapper can help use repos?

**Review by jerrymarino — APPROVED** [review]
@luispadron I think it's time to land this and forward fix. For the rules_apple 3.x update - I will rebase the changes once we land it.

**Review by jszumski — APPROVED** [review]

File: rules/module_extensions.bzl
Comment (justinseanmartin) [inline rules/module_extensions.bzl:15]: FYI - This confused me, because I thought "load_bzlmod_dependencies" means "dependencies that bzlmod needs" rather than "dependencies that bzlmod has already supplied". Just putting this here in case anyone else stumbles across this, and as always, naming is hard.

Comment (luispadron) [reply rules/module_extensions.bzl:15]: I had to update it the other day and I felt the same way lol

Happy to change this to something more clear

**Pull request merged**
